### ISSUE TITLE
server or clang preamble out of date after code refactoring with new header files?

### ISSUE THREAD

**Issue by fspeech** [Sarcasm/irony-mode#134]
After I refactor some code into new header files they seem to get lost from the completer. Restarting Emacs restores the completer to a good state. I suspect restarting the server might as well. This may be a clang issue or it may be a known issue for Irony, in which case feel free to close this.

Regarding issue #131 GNU is taking their time releasing a Windows binary of Emacs 24.4. Until that is done I suspect Windows usage may be a problem for many users. OTOH my patched version of Irony is doing well for me with 24.3 so I am not in a hurry to make further progress on #131 either.

**Comment by fspeech** [comment]
Cool. It may be worthwhile to put this to the clang devs in case this is a known issue. While I don't doubt that they would prefer a clean test case designed to isolate the bug location, the effort to do that would be wasted if it is a known issue (at least to some devs there). Of course they likely don't know much about Irony  so it would be nice if the report focus on how libclang is involved.

**Comment by Sarcasm** [comment]
Okay, so just for the record, I have made progress very slowly. I put this on the cfe-dev ML but without much feedback, and trying to reproduce the issue with a minimal libclang program doesn't seem to work. So maybe something is wrong in irony-mode usage of libclang, I will have to check the code.

**Comment by fspeech** [comment]
Thanks for the update. Does ycm use libclang the same way as irony? That
may provide some clue?

**Comment by pfilipovv** [comment]
Yes, this preamble bug is really annoying. I confirm I came accross it too. Many times :)

**Comment by fspeech** [comment]
Greetings! I have not been actively working on C++ code for a while but I am now coming back to it. I have refreshed my Irony and clang install on MinGW and so far things appear to be working smoothly as far as completion is concerned. However I have not been heavily editing the header sections so this experience is limited. Is this preamble corruption issue still a real problem?

**Comment by Sarcasm** [comment]
I think it is still an issue as I haven't debugged it further for a long time, maybe you could try to reproduce it following the same steps I describe here: http://lists.cs.uiuc.edu/pipermail/cfe-dev/2015-January/040691.html and here https://github.com/Sarcasm/irony-mode/issues/134#issuecomment-68412195

If it's still present I will be curious to try it with my local changes.

**Comment by fspeech** [comment]
Yes I can still reproduce the issue exactly as you described in your links. Bummer! My clang is at 3.6.2 (mingw64/msys2 build).

One interesting thing I noticed is that flycheck never seems to suffer the same false syntax errors. I assume flycheck and company use the same irony server? But I guess they issue different commands to the server and this problem is only triggered in the completion command.

**Comment by fspeech** [comment]
I wonder if you have looked at [how rtags is doing completion](https://github.com/Andersbakken/rtags/blob/master/src/CompletionThread.cpp#L261-L263), in particular they seem to set ```CXTranslationUnit_SkipFunctionBodies``` which is not found in ```irony```.

**Comment by Sarcasm** [comment]
Should that solve the preamble issue or improve speed?

Did you try it? I remember trying it in the past but I am now fine with how completion works so I don't change it.

**Comment by yangyingchao** [comment]
I can still reproduce this issue with clang-3.9.

And I modified TUManager.cpp, in TUManager::getOrCreateTU(), reparse TU every time for cached TU before returning it to the caller, and this issue went away.

It may slow down the completion, but at least it works...

**Comment by Nephyrin** [comment]
I've been seeing this, or at least a very similar issue, often.  Irony reports many nonsensical errors in the file that seem to indicate it is missing information about headers, such as knowing of members of classes, but not of their base classes.

Of note, I'm running both rtags and Irony, which may be somehow clashing with their un-prefixed /tmp/ filenames?

One workaround I've found is to add a line of non-includes to the start of the affected files:
```c++
bool gIrony = true;

#include <...>
...
```

This works with anything before the #include, even errors, such as `non_existant_symbol;`.  Is it possible to have irony prepend a no-op line prior to the files it is submitting to clang as a workaround?  Certainly not pretty, but this issue is making irony far less useful -- right now 2/3 files I have open are mis-parsed unless I apply said workaround.  I certainly don't have the time to dig deep into the clang code :(

This seems to have become much more common with clang 5.0.0, being near constant for me now, but that may be coincidental.
